### ISSUE TITLE
[Feature Request] attach()

### ISSUE THREAD

**Issue by Kaixhin** [pytorch/pytorch#2203]
Some applications of neural networks could benefit from the addition of an `.attach()` method to complement `.detach()`. Example:

```
x -> f() -> g() -> h() -> j() -> y -> [loss1(y, t1), loss2(y, t2), loss3(y, t3)]
```

Where we want to backprop `loss1` to get gradients only up to `h()`, manipulate these to create another loss based on these gradients, and then backprop these all the way (alongside other normal losses). This is essentially what is done in the "efficient" trust region method of [Sample Efficient Actor-Critic with Experience Replay](https://arxiv.org/abs/1611.01224). There is an example of this being achieved [in Chainer](https://github.com/chainer/chainerrl/blob/master/chainerrl/agents/acer.py#L193-L200).

The current solution would be to `detach()` after `g()`, keep the original and copy going through the network and losses, do the above on the detached copy, and then pass the gradients/new loss to the original. Achieving what Chainer does is tricky as Pytorch does not store intermediate `.grad`s, but it could perhaps be achieved with some manual work from the user to "ready" parts of the computation graph for this.

**Comment by vadimkantorov** [comment]
This would also resolve https://github.com/pytorch/pytorch/issues/1734

**Comment by Kaixhin** [comment]
Ah sorry didn't see that. Closing as previously discussed in https://github.com/pytorch/pytorch/issues/1734.
